doc/acl/Everflow-test-plan.md

```diff
@@ -0,0 +1,1213 @@
+
+Combining these two enhancements, there are 4 scenarios for everflow.
+1. ACL table of type `MIRROR` have `stage` field ignored or set to "ingress". Action type of ACL_RULE is `MIRROR_ACTION` or `MIRROR_INGRESS_ACTION`.
+2. ACL table of type `MIRROR` have `stage` field ignored or set to "ingress". Action type of ACL_RULE is `MIRROR_EGRESS_ACTION`.
+3. ACL table of type `MIRROR` have `stage` field set to "egress". Action type of ACL_RULE is `MIRROR_ACTION` or `MIRROR_INGRESS_ACTION`.
+4. ACL table of type `MIRROR` have `stage` field set to "egress". Action type of ACL_RULE is `MIRROR_EGRESS_ACTION`.
+
+Expected behaviors for the combinations:
+
+| -                                  | ACL table stage: ingress                                 | ACL table stage: egress                                |
+| ---------------------------------- | -------------------------------------------------------- | ------------------------------------------------------ |
+| Action type: MIRROR_INGRESS_ACTION | Ingress packets hit ACL rules, mirrored at ingress stage | Not applicable                                         |
+| Action type: MIRROR_EGRESS_ACTION  | Ingress packets hit ACL rules, mirrored at egress stage  | Egress packets hit ACL rules, mirrored at egress stage |
+
+Since not all the combinations are supported by all vendors, the enhancement also added ACL capability detection. The supported ACL action types at different stage are detected and stored in redis. The below is an example of showing detected capabilities:
+```
+$ redis-cli -n 6 hgetall 'SWITCH_CAPABILITY|switch'
+ 1) "MIRROR"
+ 2) "true"
+ 3) "MIRRORV6"
+ 4) "true"
+ 5) "ACL_ACTIONS|INGRESS"
+ 6) "PACKET_ACTION,REDIRECT_ACTION,MIRROR_INGRESS_ACTION"
+ 7) "ACL_ACTIONS|EGRESS"
+ 8) "PACKET_ACTION,MIRROR_EGRESS_ACTION"
+ 9) "ACL_ACTION|PACKET_ACTION"
+1)  "DROP,FORWARD"
+```
+In the above example output, two everflow combinations are supported on the platform being checked:
+* INGRESS stage, MIRROR_INGRESS_ACTION
+* EGRESS stage, MIRROR_EGRESS_ACTION
+
+This test plan needs to be extended to cover both the existing everflow function and the newly added capabilities.
+
+#### Some existing areas not covered by the existing scripts
+
+##### ACL rule for matching "IN_PORTS"
+Now the SONiC ACL rules support matching "IN_PORTS". New ACL rule for matching "IN_PORTS" need to be added and covered.
+
+#### ACL rule for matching ICMP type and code
+The SONiC ACL rules also support matching ICMP type and code. New ACL rules for matching ICMP type and code need to be added and covered. The acl-loader utility does not support ICMP type and code yet. The sonic-cfggen tool will be used for directly loading such ACL rules to config_db.
+
+##### IPv6 everflow
+The existing scripts only covered IPv6. IPv6 is also supported by SONiC now. The scripts need to be extended to cover IPv6 everflow too. To cover IPv6:
+* Everflow ACL table of type "MIRRORV6" needs to be defined and loaded during testing.
```

Comment:
Are both the MIRROR and MIRRORv6 ACL Tables created together?

Some devices would be able to fit both IPv4 and IPv6 rules in the same ACL table; in this scenario creating two ACL tables for mirroring would result in wasting the h/w resources. Is there a way we can define based on the platform whether to have separate mirror ACL tables for v4 and v6?

Reply:
@rck-innovium For simplicity, the MIRROR and MIRRORv6 ACL tables can be created together in configuration/setup stage of the testing. 

It's for testing everflow, as long as h/w resources are enough, it is not a issue. Why would we need to save h/w resources when we have enough of them? It's not for production.

If on platforms which hardware resources are not enough to do the testing, the script can create ACL table for testing each scenario. Then cleanup the ACL rules and tables after testing the scenario is completed.